modyn/storage/src/internal/grpc/storage_service_impl.cpp

```diff
@@ -0,0 +1,596 @@
+    soci::rowset<soci::row> rs =  // NOLINT misc-const-correctness
+        (session.prepare << "SELECT sample_id, label FROM samples WHERE file_id = :file_id", soci::use(file_id));
+    T response;
+    for (auto& row : rs) {
+      response.add_keys(row.get<long long>(0));    // NOLINT google-runtime-int
+      response.add_labels(row.get<long long>(1));  // NOLINT google-runtime-int
+      if (response.keys_size() == sample_batch_size_) {
+        writer->Write(response);
+        response.Clear();
+      }
+    }
+
+    if (response.keys_size() > 0) {
+      writer->Write(response);
+    }
+  }
+}
+
+template <typename T>
+void StorageServiceImpl::send_samples_asynchronous_retrieval(::grpc::ServerWriter<T>* writer, int64_t file_id,
+                                                             soci::session& session) {
+  const int64_t number_of_samples = get_number_of_samples_in_file(file_id, session);
+  if (number_of_samples <= sample_batch_size_) {
+    // If the number of samples is less than the sample batch size, retrieve all of the samples in one go.
+    soci::rowset<soci::row> rs =  // NOLINT misc-const-correctness
+        (session.prepare << "SELECT sample_id, label FROM samples WHERE file_id = :file_id", soci::use(file_id));
+    T response;
+    for (auto& row : rs) {
+      response.add_keys(row.get<long long>(0));    // NOLINT google-runtime-int
+      response.add_labels(row.get<long long>(1));  // NOLINT google-runtime-int
+    }
+    writer->Write(response);
+  } else {
+    // If the number of samples is greater than the sample batch size, retrieve the samples in batches of size
+    // sample_batch_size_. The batches are retrieved asynchronously and the futures are stored in a queue. When the
+    // queue is full, the first future is waited for and the response is sent to the client. This is repeated until all
+    // of the futures have been waited for.
+    std::queue<std::future<SampleData>> sample_ids_futures_queue;
+
+    for (int64_t i = 0; i < number_of_samples; i += sample_batch_size_) {
+      if (static_cast<int64_t>(sample_ids_futures_queue.size()) == retrieval_threads_) {
+        // The queue is full, wait for the first future to finish and send the response.
+        T response;
+
+        SampleData sample_data = sample_ids_futures_queue.front().get();
+        sample_ids_futures_queue.pop();
+
+        for (size_t i = 0; i < sample_data.ids.size(); i++) {
+          response.add_keys(sample_data.ids[i]);
+          response.add_labels(sample_data.labels[i]);
+        }
+
+        writer->Write(response);
+      }
+
+      // Start a new future to retrieve the next batch of samples.
+      std::future<SampleData> sample_ids_future =
+          std::async(std::launch::async, get_sample_subset, file_id, i, i + sample_batch_size_ - 1,  // NOLINT
+                     std::ref(storage_database_connection_));
+      sample_ids_futures_queue.push(std::move(sample_ids_future));
+    }
+
```

Comment:
When are the future/async functions actually executed?

Reply:
As soon as we do `std::async(std::launch::async` the future should start to run and finish in some time in the queue where we then `get` it from the queue.